stdlib/public/Darwin/Foundation/Data.swift

```diff
@@ -1084,79 +1039,90 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
             storage.copyBytes(to: pointer, from: range)
         }
 
-        @inlinable
-        var hashValue: Int {
-            let hashRange = startIndex..<Swift.min(startIndex + 80, endIndex)
-            return storage.withUnsafeBytes(in: hashRange) {
-                return Int(bitPattern: CFHashBytes(UnsafeMutablePointer(mutating: $0.baseAddress?.assumingMemoryBound(to: UInt8.self)), $0.count))
+        @inlinable @inline(__always) // This should always be inlined into _Representation.hash(into:).
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(count)
+
+            // At most, hash the first 80 bytes of this data.
+            let range = startIndex ..< Swift.min(startIndex + 80, endIndex)
+            storage.withUnsafeBytes(in: range) {
+                hasher.combine(bytes: $0)
```

Comment:
We should simply feed the full contents to the hasher here rather than any arbitrary subset. (Equality isn't based on the first 80 bytes, so neither should hashing.)

Reply:
Hashing should not be O(N)

Reply:
This is not really a concern -- equality checks are O(n) anyway, and any successful hash table lookup will involve one full-size comparison.

Full hashing may slow down negative lookups for *certain* inputs, but we're fine with that. All collections in the stdlib (including `String`, `Array`, `Set`, `Dictionary` etc.) hash their full contents.

The problem with such partial hashes is that they negate the advantages of randomly seeded hashing by introducing a trivial way to bypass it.

Reply:
Sure, but for collections which hash, hashing happens far, far more often than equality checks, and negative lookups are for more common than positive lookups. `Data` longer than 80 bytes are very rarely equal in those first 80 bytes, and differ after that.

If this is a blocking concern then I'm willing to change it, but from our experience with how `Data` is used, this is a reasonable middle ground between efficiency and full-size comparison.

Reply:
Not sure if this is relevant, but the trivial way to bypass robust hashing @lorentey mentioned can be a vector for DOS attacks, which seems quite likely, considering `Data` is often used to hold bytes that come from outside. Might this issue be viewed as security concern?

Reply:
TBH I am not certain how often Data or even NSData is used as a key in a Dictionary or a member of a Set (or any other hashed entry), but in reality my guess is it is very uncommon; Strings or integer types are most likely the common case by a great margin.

 The distribution of real-world usage cases of Dictionaries are overwhelmingly 0 and 1 element dictionaries, which I am not certain if Dictionary makes any fast paths for those cases. So any initial insertion would be 2 hashes and an equality. In terms of big-o notation it is both O(N) where N is the length of the Data, but making the hashes O(N) would be 3x the passes. 

NSData/NSMutableData/dispatch_data_t in real world samples often are of a few different common sizes: most common is empty (hence our refactor for the special case for empty datas), the second most common group is just under 20 bytes or so, then there are common storages of 1024 bytes and 4096 bytes, and then a distribution of sizes like 1mb and beyond. Those higher points are a few standard deviations from the mean. 

Part of the reason why NSData hashes similarly to the original hash method was done was specifically due to analysis of how these types are used. Since there is a distinct impact potential of large payloads being sourced from external sources it seems like a remote attack could be launched if there wasn't some sort of cap.

Reply:
The frequency of use argument cuts both ways. If `Data` isn't often used as a `Dictionary` key, then it's not worth overthinking its hashing -- `String` hashes all its contents, so `Data` should do the same. This is a sane default that leads to consistent, unsurprising behavior *even in the face of actively hostile input*.

I expect the vast majority of everyday `Data` keys used in practice are well below 80 bytes, so in regular use this wouldn't matter at all. (Alas, I have zero actual data to support or disprove this statement, beyond my own limited experience.) 

What I'm specifically worried about is cases where a `Dictionary<Data, Foo>` is fed hostile data that's slightly above this arbitrary and undocumented limit. I'm concerned that this is an open invitation for abuse.

Reply:
Isn't the security argument for random-seed hashing about preventing abusing hash speeds by causing expensive hashes? How does that square with making hashing always be O(n)?

Reply:
Random seeding only works as long as the hash function is fed *exactly* the same bits as are compared during equality checking. Components that are compared but not hashed provide a trivial way to completely bypass seeding.

My position is that hashing approximates half of a successful equality check, so it's okay for `hash(into:)` to have performance proportional to `==`. This is definitely easier to explain and to prepare for than a hash table that wildly misbehaves for certain cleverly constructed input sets.

I agree that if we can guarantee certain things about the shape of the input, then we can get away with partial hashing. But these are general-purpose data structures -- we don't know anything about the sorts of data they will contain.

Reply:
One interesting alternative idea for optimizing hashing is to switch to progressively stronger hash functions based on the size of the *hash table*. For a `Dictionary` below, say, a hundred keys, it may indeed be a good idea to cut hashing off at 80 bytes. Luckily, we can still do this, as long as we keep these definitions non-inlinable -- so that in some future release we'll be still able to optionally switch off full hashing when some new `Hasher` property indicates it's safe to do so.

Reply:
@lorentey, what do you mean by progressively stronger hash functions here?

----

@parkera 
> Isn't the security argument for random-seed hashing about preventing abusing hash speeds by causing expensive hashes? How does that square with making hashing always be O(n)?

I think these are 2 different Big-O complexities we are talking about. The O(1) is about lookup the performance of the hashed collections. Computation of the hash value for a given object is O(n) — proportional to the size of the size of the fields significant for equality comparison.

I think the advice by Josh Bloch from Effective Java about [computing hashCode](http://bit.ly/2CeCQPz) is still relevant in Swift:
> **Do not be tempted to exclude significant fields from the hash code computation to improve performance.** While the resulting hash function may run faster, its poor quality may degrade hash tables' performance to the point where they become unusable. In particular, the hash function may be confronted with a large collection of instances that differ mainly in regions you've chosen to ignore. If this happens, the hash function will map all these instances to a few hash codes, and programs that should tun in linear time will instead run in quadratic time.

See also [Guarding Against Hash-Flooding](https://nshipster.com/hashable/#guarding-against-hash-flooding)… I think the point about `stdlib` and `Foundation` provided types to have safe implementation stands. If we are that concerned with performance of `Data` hashing, we should investigate the further advice from Bloch:

> If a class is immutable and the cost of computing the hash code is significant, you might consider caching the hash code in the object rather than recalculating it each time it is requested, If you believe that most objects of this type will be used as hash keys, then you should calculate the hash code when the instance is created. Otherwise, you might choose to *lazily initialize* the hash code the first time `hashCode` is invoked.

Reply:
Hash collisions are generally only an issue for relatively large hash tables; most Set/Dictionary instances are too small to care. I agree full hashing is overkill for these. There is room for future optimizations to allow small hash tables to switch to a less complicated hash operation.

However, it is not up to `Hashable` types to decide if it's appropriate to use incomplete hashing. They simply don't have enough information about the context in which they are hashed to make an informed decision.

`Hashable` makes no explicit requirements about the performance of hashing itself. The design assumption for `Set` and `Dictionary` is that the time cost of hashing is proportional to a successful equality check. A `Data.hash(into:)` implementation that takes time proportional to the length of its contents is perfectly fine.

Reply:
Let me know if I understood your point correctly: The `Hasher` called to `combine(bytes: UnsafeRawBufferPointer)` might decide to subsample the buffer when computing the hash, but the `Data`, should be calling it with full range from the storage?